sui_core/src/full_node/full_node_state.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use std::{
+    collections::HashSet,
+    sync::{atomic::AtomicU64, Arc},
+};
+use sui_config::genesis::Genesis;
+
+use crate::{
+    authority::{AuthorityTemporaryStore, ReplicaStore},
+    gateway_state::{gateway_responses::TransactionEffectsResponse, GatewayTxSeqNumber},
+};
+use move_binary_format::CompiledModule;
+use move_vm_runtime::{move_vm::MoveVM, native_functions::NativeFunctionTable};
+use sui_adapter::adapter;
+use sui_types::{
+    base_types::{ObjectID, TransactionDigest, TxContext},
+    committee::Committee as SuiCommittee,
+    error::{SuiError, SuiResult},
+    fp_ensure,
+    gas::SuiGasStatus,
+    messages::Transaction,
+    object::Object,
+    MOVE_STDLIB_ADDRESS, SUI_FRAMEWORK_ADDRESS,
+};
+use tracing::debug;
+
+const MAX_TX_RANGE_SIZE: u64 = 4096;
+
+pub struct FullNodeState {
+    pub store: Arc<ReplicaStore>,
+    pub committee: SuiCommittee,
+    pub next_tx_seq_number: AtomicU64,
+
+    /// Move native functions that are available to invoke
+    _native_functions: NativeFunctionTable,
+    /// Will be used for local exec in future
+    _move_vm: Arc<MoveVM>,
+}
+
+impl FullNodeState {
+    pub async fn new_without_genesis(
+        committee: SuiCommittee,
+        store: Arc<ReplicaStore>,
+    ) -> Result<Self, SuiError> {
+        let native_functions =
+            sui_framework::natives::all_natives(MOVE_STDLIB_ADDRESS, SUI_FRAMEWORK_ADDRESS);
+        let next_tx_seq_number = AtomicU64::new(store.next_sequence_number()?);
+        Ok(Self {
+            committee,
+            store,
+            _native_functions: native_functions.clone(),
+            _move_vm: Arc::new(
+                adapter::new_move_vm(native_functions)
+                    .expect("We defined natives to not fail here"),
+            ),
+            next_tx_seq_number,
+        })
+    }
+
+    pub async fn new_with_genesis(
+        committee: SuiCommittee,
+        store: Arc<ReplicaStore>,
+        genesis: &Genesis,
+    ) -> Result<Self, SuiError> {
+        let state = Self::new_without_genesis(committee, store.clone()).await?;
+
+        // Only initialize an empty database.
+        if store
+            .database_is_empty()
+            .expect("Database read should not fail.")
+        {
+            let mut genesis_ctx = genesis.genesis_ctx().to_owned();
+            for genesis_modules in genesis.modules() {
+                state
+                    .store_package_and_init_modules_for_genesis(
+                        &mut genesis_ctx,
+                        genesis_modules.to_owned(),
+                    )
+                    .await
+                    .expect("We expect publishing the Genesis packages to not fail");
```

Comment:
Missing importing of objects?
```
            state
                .insert_genesis_objects_bulk_unsafe(&genesis.objects().iter().collect::<Vec<_>>())
                .await;
```

I wonder if we could merge the authority and fullnode storages so we don't have to replicate this logic here.

Reply:
Ah yes. Rebase mistake